pkg/kubelet/kubelet_getters.go

```diff
@@ -234,18 +232,9 @@ func (kl *Kubelet) getRuntime() kubecontainer.Runtime {
 
 // GetNode returns the node info for the configured node name of this Kubelet.
 func (kl *Kubelet) GetNode() (*v1.Node, error) {
-	if kl.kubeClient == nil {
+	if kl.kubeClient == nil || !kl.nodeHasSynced() {
 		return kl.initialNode(context.TODO())
```

Comment:
/hold

quickly returning `initialNode()` before the node informer has synced is not correct... doesn't that mean existing pods that are currently running could be evicted on a kubelet restart if their node selectors did not match the kubelet's self-computed initial labels?

Reply:
this seems like it is reintroducing the correctness problem https://github.com/kubernetes/kubernetes/pull/94087 fixed

Reply:
> doesn't that mean existing pods that are currently running could be evicted on a kubelet restart if their node selectors did not match the kubelet's self-computed initial labels?

i did not see it in practice after a kubelet restart, but could happen in theory.

> this seems like it is reintroducing the correctness problem #94087 fixed

as mentioned on https://github.com/kubernetes/kubernetes/pull/94087, the wait for sync can also happen once on startup. 
but IIUC this means the kubelet has to start the static pods and wait for sync, without calling on GetNode().

will experiment more, but https://github.com/kubernetes/kubernetes/pull/94087 has to be iterated on.

Reply:
We meet #94087 sometimes and finally, we use a logic that is similar to comments here https://github.com/kubernetes/kubernetes/issues/92067#issuecomment-643711902 to avoid re-scheduling pods after kubelet restart.

Reply:
> We meet #94087 sometimes and finally, we use a logic that is similar to comments here #92067 (comment) to avoid re-scheduling pods after kubelet restart. It works well in our product env.

looking at the code in the comment, it's returning the initial node if `kl.nodeLister.Get()` fails. this current PR is returning `kl.nodeLister.Get()` only after the informer sync. so this is similar.

based on your comment and the linked comment, does that confirm that returning the initial node mitigates the problem?

> I prefer to wait in getter only.

in general, for a system that is not only kubernetes, i would say i prefer waiting for a module to be ready before using it in getters. but i don't mind fixing the problem one way or the other, because the users will soon start reporting the timing issue.

Reply:
What if we have an alternative `nodeLister.GetForStaticPods()` method that doesn't have the poll? That way, static pods are created ASAP with a best-effort Node.

Reply:
part of the slowness problem here is that GetNode and getNodeAnyway are called so many times that it just creates a lot of polls.

example patch:

<details>

```
diff --git a/pkg/kubelet/kubelet_getters.go b/pkg/kubelet/kubelet_getters.go
index 32e80cc863a..34da817ab3a 100644
--- a/pkg/kubelet/kubelet_getters.go
+++ b/pkg/kubelet/kubelet_getters.go
@@ -22,6 +22,7 @@ import (
        "io/ioutil"
        "net"
        "path/filepath"
+       "runtime"
        "time"
 
        cadvisorapiv1 "github.com/google/cadvisor/info/v1"
@@ -238,10 +239,22 @@ func (kl *Kubelet) GetNode() (*v1.Node, error) {
                return kl.initialNode(context.TODO())
        }
        // if we have a valid kube client, we wait for initial lister to sync
+
        if !kl.nodeHasSynced() {
+
+               pc, fn, line, _ := runtime.Caller(1)
+               klog.Infof("------------ %v %v %v", runtime.FuncForPC(pc).Name(), fn, line)
+               pc, fn, line, _ = runtime.Caller(2)
+               klog.Infof("----------------- %v %v %v", runtime.FuncForPC(pc).Name(), fn, line)
+
+               t := time.Now().UnixNano()
+
+               klog.Info("----- start poll", t)
                err := wait.PollImmediate(time.Second, maxWaitForAPIServerSync, func() (bool, error) {
                        return kl.nodeHasSynced(), nil
                })
+               klog.Info("----- end poll", t)
+
                if err != nil {
                        return nil, fmt.Errorf("nodes have not yet been read at least once, cannot construct node object")
                }

```

</details>


output:

<details>

```
lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 271
фев 23 17:57:14 lubo-it kubelet[34967]: I0223 17:57:14.674992   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getHostIPsAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 300
фев 23 17:57:14 lubo-it kubelet[34967]: I0223 17:57:14.674995   34967 kubelet_getters.go:251] ----- start poll1614095834674993617
фев 23 17:57:23 lubo-it kubelet[34967]: I0223 17:57:23.395489   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet/server/stats.(*summaryProviderImpl).Get /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/summary.go 70
фев 23 17:57:23 lubo-it kubelet[34967]: I0223 17:57:23.395516   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet/eviction.(*managerImpl).synchronize /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/eviction/eviction_manager.go 253
фев 23 17:57:23 lubo-it kubelet[34967]: I0223 17:57:23.395536   34967 kubelet_getters.go:251] ----- start poll1614095843395523312
фев 23 17:57:23 lubo-it kubelet[34967]: I0223 17:57:23.611550   34967 kubelet_getters.go:255] ----- end poll1614095833609464983
фев 23 17:57:23 lubo-it kubelet[34967]: I0223 17:57:23.713866   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).fastStatusUpdateOnce /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 2271
фев 23 17:57:23 lubo-it kubelet[34967]: I0223 17:57:23.713939   34967 kubelet_getters.go:251] ----- start poll1614095843713931550
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.675760   34967 kubelet_getters.go:255] ----- end poll1614095834674898274
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.675759   34967 kubelet_getters.go:255] ----- end poll1614095834674993617
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689038   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncPod /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 1668
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689045   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop.func1 /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go 175
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689048   34967 kubelet_getters.go:251] ----- start poll1614095844689046894
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689673   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getNodeAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 271
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689678   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet/lifecycle.(*predicateAdmitHandler).Admit /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/lifecycle/predicate.go 63
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689681   34967 kubelet_getters.go:251] ----- start poll1614095844689679754
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689714   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getNodeAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 271
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689718   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getHostIPsAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 300
фев 23 17:57:24 lubo-it kubelet[34967]: I0223 17:57:24.689721   34967 kubelet_getters.go:251] ----- start poll1614095844689719986
фев 23 17:57:33 lubo-it kubelet[34967]: I0223 17:57:33.397687   34967 kubelet_getters.go:255] ----- end poll1614095843395523312
фев 23 17:57:33 lubo-it kubelet[34967]: I0223 17:57:33.714370   34967 kubelet_getters.go:255] ----- end poll1614095843713931550
фев 23 17:57:33 lubo-it kubelet[34967]: I0223 17:57:33.815646   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).fastStatusUpdateOnce /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 2271
фев 23 17:57:33 lubo-it kubelet[34967]: I0223 17:57:33.815662   34967 kubelet_getters.go:251] ----- start poll1614095853815659396
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.690466   34967 kubelet_getters.go:255] ----- end poll1614095844689719986
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.690703   34967 kubelet_getters.go:255] ----- end poll1614095844689679754
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.690745   34967 kubelet_getters.go:255] ----- end poll1614095844689046894
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.711520   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncPod /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 1668
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.711529   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop.func1 /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go 175
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.711533   34967 kubelet_getters.go:251] ----- start poll1614095854711531429
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.718926   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getNodeAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 271
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.718933   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet/lifecycle.(*predicateAdmitHandler).Admit /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/lifecycle/predicate.go 63
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.718936   34967 kubelet_getters.go:251] ----- start poll1614095854718935029
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.718990   34967 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getNodeAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 271
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.718994   34967 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getHostIPsAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 300
фев 23 17:57:34 lubo-it kubelet[34967]: I0223 17:57:34.718997   34967 kubelet_getters.go:251] ----- start poll1614095854718995561
фев 23 17:57:47 lubo-it kubelet[35849]: I0223 17:57:47.010245   35849 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getNodeAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 271
фев 23 17:57:47 lubo-it kubelet[35849]: I0223 17:57:47.010251   35849 kubelet_getters.go:247] ----------------- k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initializeRuntimeDependentModules /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 1370
фев 23 17:57:47 lubo-it kubelet[35849]: I0223 17:57:47.010253   35849 kubelet_getters.go:251] ----- start poll1614095867010252303
фев 23 17:57:47 lubo-it kubelet[35849]: I0223 17:57:47.044750   35849 kubelet_getters.go:245] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).fastStatusUpdateOnce /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 2271
фев 23 17:57:47 lubo-it kubelet[35849]: I0223 17:57:47.044759   35849 kubelet_getters.go:251] ----- start poll1614095867044757891
фев 23 17:58:16 lubo-it kubelet[37741]: I0223 17:58:16.913919   37741 kubelet_getters.go:246] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).getNodeAnyWay /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go 272
фев 23 17:58:16 lubo-it kubelet[37741]: I0223 17:58:16.913925   37741 kubelet_getters.go:248] ----------------- k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initializeRuntimeDependentModules /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 1370
фев 23 17:58:16 lubo-it kubelet[37741]: I0223 17:58:16.913929   37741 kubelet_getters.go:252] ----- start poll1614095896913927359
фев 23 17:58:16 lubo-it kubelet[37741]: I0223 17:58:16.958365   37741 kubelet_getters.go:246] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).fastStatusUpdateOnce /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 2271
фев 23 17:58:16 lubo-it kubelet[37741]: I0223 17:58:16.958388   37741 kubelet_getters.go:252] ----- start poll1614095896958386116
фев 23 17:58:26 lubo-it kubelet[37741]: I0223 17:58:26.915966   37741 kubelet_getters.go:256] ----- end poll1614095896913927359
фев 23 17:58:26 lubo-it kubelet[37741]: I0223 17:58:26.948723   37741 kubelet_getters.go:246] ------------ k8s.io/kubernetes/pkg/kubelet/server/stats.(*summaryProviderImpl).Get /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/summary.go 70
фев 23 17:58:26 lubo-it kubelet[37741]: I0223 17:58:26.948727   37741 kubelet_getters.go:248] ----------------- k8s.io/kubernetes/pkg/kubelet/eviction.(*managerImpl).synchronize /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/eviction/eviction_manager.go 253
фев 23 17:58:26 lubo-it kubelet[37741]: I0223 17:58:26.948730   37741 kubelet_getters.go:252] ----- start poll1614095906948729209
фев 23 17:58:26 lubo-it kubelet[37741]: I0223 17:58:26.961188   37741 kubelet_getters.go:256] ----- end poll1614095896958386116
фев 23 17:58:27 lubo-it kubelet[37741]: I0223 17:58:27.061588   37741 kubelet_getters.go:246] ------------ k8s.io/kubernetes/pkg/kubelet.(*Kubelet).fastStatusUpdateOnce /home/lubo/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go 2271
фев 23 17:58:27 lubo-it kubelet[37741]: I0223 17:58:27.061598   37741 kubelet_getters.go:252] ----- start poll1614095907061596323

```

</details>

Reply:
https://github.com/kubernetes/kubernetes/pull/99336#issuecomment-784473647